packages/core/src/browser/label-provider.ts

```diff
@@ -88,10 +100,67 @@ export class DefaultUriLabelProviderContribution implements LabelProviderContrib
 @injectable()
 export class LabelProvider {
 
+    protected readonly onDidChangeEmitter = new Emitter<DidChangeLabelEvent>();
+
     constructor(
         @inject(ContributionProvider) @named(LabelProviderContribution)
         protected readonly contributionProvider: ContributionProvider<LabelProviderContribution>
-    ) { }
+    ) {
+        const contributions = this.contributionProvider.getContributions();
+        for (const contribution of contributions) {
+            if (contribution.onDidChange) {
+                contribution.onDidChange(event => {
+                    const affects = (uri: URI) => this.affects(uri, event, contribution);
+                    this.onDidChangeEmitter.fire({ affects });
+                });
+            }
+        }
+    }
+
+    /**
+     * When the given event occurs, determine if the given URI could in any
+     * way be affected.
+     *
+     * If the event directly indicates that it affects the URI then of course we
+     * return `true`.  However there may be label provider contributions that delegate
+     * back to the label provider.  These contributors do not, and should not, listen for
+     * label provider events because that would cause infinite recursion.
+     *
+     * @param uri
+     * @param event
+     */
+    protected affects(element: URI, event: DidChangeLabelEvent, originatingContribution: LabelProviderContribution): boolean {
+
+        const contribs = this.findContribution(element);
+        const possibleContribsWithDups = [
```

Comment:
Why cannot we just iterate `contribs`? What would be the point to have a label provider contribution doing nothing? Also does not it break priority order applied in `findContribution`?

Reply:
We don't want to process all contributions, just the one that was actually used to obtain the three label properties.  As a contribution can implement only one or two of the three property getters, a different contribution may be used for each property.  So there could be anywhere from one to three contributions that were actually used to get a property. 

'contribs' is all contributions that 'can' process element, in priority order.
'possibleContribsWithDups' contains, in the first element, the actual contribution that would be called to get the icon.  This matches the code in getIcon.  Similarly the second element for the 'name' property and the third element for the 'longName' property.
'possibleContribsWithDups' contains the contributions that are actually called to get one or more of the three properties.  So these are the only contributions that we need ask.